### PR TITLE
Do not silently ignore zone close errors

### PR DESCRIPTION
**Pull request by metaspace**
Suggestion for fix fix for #68

### CODE REVIEW

**Comment by metaspace** [comment]
@yhr Added a flag that is checked during zone allocation. If any unreported errors were posted, the error is reported by next call to AllocateZone.

File: fs/zbd_zenfs.h [hunk 1]
```diff
@@ -79,6 +81,8 @@ class ZonedBlockDevice {
   std::atomic<long> active_io_zones_;
   std::atomic<long> open_io_zones_;
   std::condition_variable zone_resources_;
+  std::mutex zone_close_mutex_;
+  boost::optional<IOStatus> zone_close_error_;
```
Comment (yhr) [inline fs/zbd_zenfs.h:85]: Why do we need optional here? why not just set it to ok?

File: fs/io_zenfs.cc [hunk 2]
```diff
@@ -220,15 +220,22 @@ ZoneFile::~ZoneFile() {
     zone->used_capacity_ -= (*e)->length_;
     delete *e;
   }
-  CloseWR();
+  IOStatus status = CloseWR();
```
Comment (yhr) [inline fs/io_zenfs.cc:223]: We use "s" for status everywhere else, so please use it here as well (and in the other places you add io status variables)

File: fs/zbd_zenfs.cc [hunk 3]
```diff
@@ -467,13 +477,23 @@ void ZonedBlockDevice::ResetUnusedIOZones() {
   }
 }
 
-Zone *ZonedBlockDevice::AllocateZone(Env::WriteLifeTimeHint file_lifetime) {
+IOStatus ZonedBlockDevice::AllocateZone(Env::WriteLifeTimeHint file_lifetime,
+                                        Zone **out_zone) {
```
Comment (yhr) [inline fs/zbd_zenfs.cc:481]: "out_zone" -> allocated_zone

File: fs/zbd_zenfs.h (hunk 1 above)
Comment (yhr) [inline fs/zbd_zenfs.h:85]: Let's make this "zone_close_error_" a generic "zone_management_status" in stead so we can use it for other stuff that goes wrong in the future. zone_management_status should be initialized to ok.

File: fs/zbd_zenfs.cc [hunk 4]
```diff
@@ -467,13 +477,23 @@ void ZonedBlockDevice::ResetUnusedIOZones() {
...
   Zone *finish_victim = nullptr;
   unsigned int best_diff = LIFETIME_DIFF_NOT_GOOD;
   int new_zone = 0;
   Status s;
 
+  *out_zone = nullptr;
+
+  {
+    std::lock_guard<std::mutex> lock(this->zone_close_mutex_);
+    if (this->zone_close_error_ != boost::none) {
```
Comment (yhr) [inline fs/zbd_zenfs.cc:492]: you don't need this->, and you can just use if(!zone_close_error->ok()) which is a lot easier to read and understand

**Review by yhr — CHANGES_REQUESTED** [review]

File: fs/zbd_zenfs.h (hunk 1 above)
Comment (metaspace) [reply fs/zbd_zenfs.h:85]: I think the intention of the code carries more clearly when using an option.

File: fs/io_zenfs.cc (hunk 2 above)
Comment (metaspace) [reply fs/io_zenfs.cc:223]: Just trying to introduce some best practice. Using descriptive names for variables carries intent more clearly for the reader.

File: fs/zbd_zenfs.cc (hunk 3 above)
Comment (metaspace) [reply fs/zbd_zenfs.cc:481]: Prefixing pointer arguments used for returning allocated items with `out_` conveys intention to the reader. How about `out_allocated_zone` ?

File: fs/zbd_zenfs.cc (hunk 4 above)
Comment (metaspace) [reply fs/zbd_zenfs.cc:492]: Using the `this` pointer when accessing class members makes the code easier to read, as it is easier to see if a stack local variable or a class member that is referenced.

Comment (yhr) [reply fs/zbd_zenfs.cc:492]: The underscore indicates that it is a class member, and we don't use this-> anywhere else (and we're not converting the rest of the code).  / code style dictator

File: fs/zbd_zenfs.cc (hunk 3 above)
Comment (yhr) [reply fs/zbd_zenfs.cc:481]: that would be fine

File: fs/io_zenfs.cc (hunk 2 above)
Comment (yhr) [reply fs/io_zenfs.cc:223]: s is fine and used everywhere else

File: fs/zbd_zenfs.h (hunk 1 above)
Comment (yhr) [reply fs/zbd_zenfs.h:85]: We want the status to be ok per default

**Comment by metaspace** [comment]
@yhr rebased on master and addressed your comments :)

**Comment by metaspace** [comment]
Changed commit messages to match style.

**Review by yhr — APPROVED** [review]

**Pull request merged**